internal/backend/smb/conpool.go

```diff
@@ -0,0 +1,243 @@
+func (b *Backend) removeSession() {
+	atomic.AddInt32(&b.sessions, -1)
+}
+
+// getSessions shows whether there are any sessions in use
+func (b *Backend) getSessions() int32 {
+	return atomic.LoadInt32(&b.sessions)
+}
+
+// dial starts a client connection to the given SMB server. It is a
+// convenience function that connects to the given network address,
+// initiates the SMB handshake, and then returns a session for SMB communication.
+func (b *Backend) dial(ctx context.Context, network, addr string) (*conn, error) {
+	dialer := net.Dialer{}
+	tconn, err := dialer.Dial(network, addr)
+	if err != nil {
+		return nil, err
+	}
+	var clientID [16]byte
+	if b.ClientGUID != "" {
+		copy(clientID[:], []byte(b.ClientGUID))
+	}
+
+	d := &smb2.Dialer{
+		Negotiator: smb2.Negotiator{
+			RequireMessageSigning: b.RequireMessageSigning,
+			SpecifiedDialect:      b.Dialect,
+			ClientGuid:            clientID,
+		},
+		Initiator: &smb2.NTLMInitiator{
+			User:     b.User,
+			Password: b.Password.Unwrap(),
+			Domain:   b.Domain,
+		},
+	}
+
+	session, err := d.DialContext(ctx, tconn)
+	if err != nil {
+		return nil, err
+	}
+
+	return &conn{
+		smbSession: session,
+		conn:       &tconn,
+	}, nil
+}
+
+// Open a new connection to the SMB server.
+func (b *Backend) newConnection(share string) (c *conn, err error) {
+	// As we are pooling these connections we need to decouple
+	// them from the current context
+	ctx := context.Background()
+
+	c, err = b.dial(ctx, "tcp", b.Host+":"+strconv.Itoa(b.Port))
+	if err != nil {
+		return nil, fmt.Errorf("couldn't connect SMB: %w", err)
+	}
+
+	if share != "" {
+		// mount the specified share as well if user requested
+		c.smbShare, err = c.smbSession.Mount(share)
+		if err != nil {
+			_ = c.smbSession.Logoff()
+			return nil, fmt.Errorf("couldn't initialize SMB: %w", err)
+		}
+		c.smbShare = c.smbShare.WithContext(ctx)
+	}
+
+	return c, nil
+}
+
+// Ensure the specified share is mounted or the session is unmounted
+func (c *conn) mountShare(share string) (err error) {
+	if c.shareName == share {
+		return nil
+	}
+	if c.smbShare != nil {
+		err = c.smbShare.Umount()
+		c.smbShare = nil
+	}
+	if err != nil {
+		return
+	}
+	if share != "" {
+		c.smbShare, err = c.smbSession.Mount(share)
+		if err != nil {
+			return
+		}
+	}
+	c.shareName = share
+	return nil
+}
+
+// Get a SMB connection from the pool, or open a new one
+func (b *Backend) getConnection(ctx context.Context, share string) (c *conn, err error) {
+	b.poolMu.Lock()
+	for len(b.pool) > 0 {
+		c = b.pool[0]
+		b.pool = b.pool[1:]
+		err = c.mountShare(share)
+		if err == nil {
+			break
+		}
+		debug.Log("Discarding unusable SMB connection: %v", err)
+		c = nil
+	}
+	b.poolMu.Unlock()
+	if c != nil {
+		return c, nil
+	}
+	c, err = b.newConnection(share)
+	return c, err
+}
+
+// Return a SMB connection to the pool
+//
+// It nils the pointed to connection out so it can't be reused
+func (b *Backend) putConnection(pc **conn) {
```

Comment:
Setting *pc back to nil is too easily defeated to be useful:

```
c, _ := b.getConnection(ctx, share)
c2 := c
b.putConnection(&c)
// c is nil, but c2 still refers to the conn.
```

I suggest just passing the pointer to this function. That saves some code and prevents the pointer from getting heap-allocated.

Reply:
Got it. Thank you!